File: solutions/lua/change/1/change.lua
return function(target, coins)
  assert(target >= 0, "target can't be negative")
  local changes = { [0] = {} }
  for t = 1, target do
    local best = nil
    for _, coin in ipairs(coins) do
      local prev = changes[t - coin]
      if prev then
        local candidate = { coin, table.unpack(prev) }
        if best == nil or #candidate < #best then best = candidate end
      end
    end
    changes[t] = best
  end
  return assert(changes[target], "can't make target with given coins")
end

File: solutions/clojure/change/1/src/change.clj
(ns change)

(defn issue [target coins]
  (when (neg? target)
    (throw (IllegalArgumentException. "target can't be negative")))

  (let [change-table
        (reduce
         (fn [change-table amount]
           (conj change-table
                 (->> coins
                      (keep (fn [coin]
                              (when-some [previous-change (get change-table (- amount coin))]
                                (conj previous-change coin))))
                      (sort-by count)
                      first)))
         [[]]
         (range 1 (inc target)))]
    (if-some [change (peek change-table)]
      (sort change)
      (throw (IllegalArgumentException.
              "can't make target with given coins")))))
